File: Session14/C++/BipartiteGraph.cpp
https://leetcode.com/problems/is-graph-bipartite/

class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {  
        
        int vertices = graph.size();
        vector<int> color(vertices, 0);
        
        for(int i=0; i<vertices; i++) {
            if(color[i]==0){
                queue<int> q;
                q.push(i);
                color[i] = 1;
    
                while(!q.empty()) {
                    int node = q.front();
                    q.pop();
                    for(int nbr:graph[node]){
                        // uncolored
                        if(color[nbr]==0) {
                            color[nbr] = -color[node];
                            q.push(nbr);
                        }
                        // same color
                        else if(color[node]==color[nbr]) {
                            return false;
                        }
                    }
                }
            }
        }
        return true;
    }
};

File: Session14/Java/BipartiteGraph.java
https://leetcode.com/problems/is-graph-bipartite/

class Solution {
    public boolean isBipartite(int[][] graph) {
        
        int len = graph.length;
        int[] colors = new int[len];
        
        for (int i=0; i<len; i++) {
            if (colors[i] == 0) {
                Queue<Integer> queue = new LinkedList<>();
                queue.add(i);
                colors[i] = 1;  
                while (!queue.isEmpty()) {
                    int node = queue.poll();
                    for (int nbr : graph[node]) {
                        // uncolored
                        if (colors[nbr] == 0) {          
                            colors[nbr] = -colors[node];  
                            queue.add(nbr);
                        } 
                        // same color
                        if(colors[nbr] == colors[node]) {
                            return false;
                        }
                    }
                }
            }
        }
        return true;
    }
}